### ISSUE TITLE
[mysql] Query can not cancel. worker process runaway

### ISSUE THREAD

**Issue by dozen** [getredash/redash#2302]
### Issue Summary

MySQL query can not cancel.

When canceling a query whose data source is MySQL, celery runaway.

### Steps to Reproduce

1. Execute time-consuming queries
example query: 
```
SELECT SLEEP(10);
```

2. Push the Cancel button. Return error `Error running query: terminated`
![image](https://user-images.githubusercontent.com/802038/36024850-5feece58-0dd4-11e8-9cea-3d2380776dec.png)

3. worker process runaway. CPU usage exhaust.
```
top - 11:01:06 up 1 day,  4:18,  0 users,  load average: 0.78, 0.45, 0.38
Tasks:   5 total,   2 running,   3 sleeping,   0 stopped,   0 zombie
%Cpu(s): 28.2 us,  3.9 sy,  0.0 ni, 67.9 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem :  6100252 total,  1165948 free,   866780 used,  4067524 buff/cache
KiB Swap:  1048572 total,  1048572 free,        0 used.  4833908 avail Mem 

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                                                  
   12 redash    20   0  314296 126064   8184 R  99.7  2.1   0:17.35 celery                                                                                   
    1 redash    20   0  289876 119336  15844 S   0.3  2.0   0:07.48 celery                                                                                   
   13 redash    20   0  299428 111200   7464 S   0.0  1.8   0:01.18 celery                                                                                   
   18 redash    20   0   36640   3104   2676 R   0.0  0.1   0:00.06 top                                                                                      
   25 redash    20   0  308192 124056   6496 S   0.0  2.0   0:00.96 celery 
```

It seems that the process will run out of control if [SIGINT is sent to the worker process](https://github.com/getredash/redash/blob/master/redash/tasks/queries.py#L205), since it is not possible to trap KeyboardInterrupt while running [MySQLdb.execute()](https://github.com/getredash/redash/blob/master/redash/query_runner/mysql.py#L144).

In order to avoid this, need to execute MySQLdb.execute () as a sub thread using threading module, trap SIGINT on the main thread, and cancel the query with the MySQL KILL statement.

Or, I think need to use a Pure Python MySQL module that allows Python to trap SIGNAL even during execute(). For example PyMySQL. Even in this case it is necessary to send the KILL syntax to MySQL (since the MySQL query will not be aborted even if the job is canceled).

I tried using threading module and tried writing a patch to avoid this problem.
I'm glad if you can use it as a reference.
https://github.com/dozen/redash/commit/d389f57251e1b39cc8dc2daee04c80ef6f0c3601#diff-180d5653c07f490439218cddb08a7dbbR153

We confirmed that this problem will be reproduced from v1.0.1 to v4.0.0.

### Technical details:

* Redash Version: 4.0.0
* Browser/OS: Google Chrome / MacOS HighSierra
* How did you install Redash: docker develoment environment

**Comment by arikfr** [comment]
Thank you for the elaborate bug report and the extra details!

I'm not too excited about the thread option, as I feel we're opening ourselves to a whole set of new problems. If using a Python implementation of the MySQL driver solves this, it feels as the safer option here.

The way cancelling queries currently works was modeled after Postgres, where if you kill the process holding the connection, PG will usually cancel the running query too. But obviously this is not applicable to all our data sources. I think that in the long term, the right solution will be to implement a dedicated `cancel` method for each query runner, that has its own logic. The cancel method will be called in a separate job, and once worked it will cause the query runner to stop. 

Some query runners won't support this at all, but it's better than the current deterministic behavior.

**Comment by jcowley** [comment]
Would love to see improved support for MySQL. We love redash, but hanging queries and the constant need to manually flush queries is makes it unreliable and frustrating for users on the business side.

**Comment by arikfr** [comment]
This is probably solved by #3003.
